rfd/0-job-worker.md

```diff
@@ -0,0 +1,236 @@
+
+Implement a prototype job worker service that provides an API to run arbitrary Linux processes.
+
+## Details
+
+### Design Approach
+
+Follow the requirements of [the challenge](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-1.md) as closely as possible and do not unnecessarily expand the scope.
+
+### Proposed API
+
+The gRPC API is defined in `jobworker.proto` and the worker library interface is defined in `worker.go`.
+
+#### Worker Library
+
+The worker library defines a simple interface with methods to start, stop, query status and get the output of a job. The implementation is responsible for starting jobs, keeping track of their output, status and exit codes.
+
+##### cgroups
+
+The library implementation will be configured at instantiation with limits that can be parsed into values for `cpu.max`, `memory.max` and `io.max`. These values will originate from runtime flags provided to the server with defaults that do not implement limits. All processes will be started in a new cgroup regardless of the values of these limits. Note that for expediency, it is assumed that cgroups v2 are in use on the system running the service.
+
+##### namespaces
+
+The library implementation will be configured such that it knows how to execute the server binary as a child process, with the proper clone and unshare flags to execute in a new pid, mount and network namespace, and call back into the library to execute the given command for the job. This library instance is obviously separate from the main server process, but due to being reexecuted with the same flags, just a different command (`child` instead of `serve`, see CLI UX below), it will retain all the necessary configuration.
+
+##### Job Execution
+
+The server will initially reexecute itself, using `exec.Command()`, as follows:
+
+```sh
+/proc/self/exe child [serve flags] -- command [args]...
+```
+
+When calling this, it will set the command's `SysProcAttr` with:
+
+```go
+syscall.SysProcAttr{
+    Cloneflags: syscall.CLONE_NEWPID | // New pid namespace
+        syscall.CLONE_NEWNS | // New mount namespace group
+        syscall.CLONE_NEWNET, // New network namespace
+    Unshareflags: syscall.CLONE_NEWNS, // Isolate process mounts from host
+}
+```
+
+In order to provide proper accounting of the job the server will internally track of the execution with:
+
+- buffers for `stdout` and `stderr`
+- process status (e.g. running, complete, error)
+- exit code
+
+Once reexecuted, the `child` command has several jobs to do. It has to remount the `/proc` filesystem, create and configure the cgroup, set the cgroup for the pid and fully replace the execution, using `syscall.Exec()`, with the job binary.
+
+##### Job Status
+
+Job status is extremely simple and only returns a status code and the exit_code. The status code is one of:
+
+- running: the job has not yet completed
+- completed: the job completed successfully
+- error: the job completed with a non-zero exit code
```

Comment:
When stopping jobs, I'll first attempt with `SIGINT` in which case, if the job exits on its own, it will have an exit code. If the job dies from that, or times out and I have to use a SIGKILL then it will not have an exit code. The exit code is optional in `JobStatusResponse`.

However, you're right, it's ambiguous. A successfully stopped (`SIGINT`) job is both completed and stopped.

I'll remove the `error` status because it can be inferred from the exit code.

Reply:
updated